Web Performance Optimization with webpack

iamakulov opened the pull request:
What's changed, or what was fixed?
- Added an article on front-end performance with webpack

**Target Live Date:** Probably January?

- [x] This has been reviewed and approved by @addyosmani 
- [x] I have run `gulp test` locally and all tests pass. (There’re a few `<script>`-related warnings, but these `<script>` blocks are code examples)
- [x] I have added the appropriate `type-something` label. (Not sure what’s this and haven’t found related info)
- [x] I've staged the site and manually verified that my content displays correctly.

**R:** @petele 

-----

I also have a few formatting-related questions:
1. For some reason, “Note:” blocks were recognized during one server launch and weren’t recognized during another. Probably that’s my local buggy installation, but IDK. Could we check this on the proper environment?

2. Locally, `{: .hide-from-toc }` wasn’t hiding “Further reading” headers from TOC (see e.g. [`decrease-frontend-size.md`](https://github.com/iamakulov/WebFundamentals/blob/webpack-article/src/content/en/fundamentals/performance/webpack/decrease-frontend-size.md). Could you please check if I’m using them properly?

3. Is there a way to enable diff-like formatting (see below) for specific code blocks? I use it in a few code examples.

```diff
- const a = 5;
+ const b = 5;
```

4. I’d like to add “Next section” arrows like in [“Optimizing Content Efficiency”](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/). How to I do this? I haven’t found any related markup, and the “Optimizing Content Efficiency” article in my local build didn’t have them.
 
![image](https://user-images.githubusercontent.com/2953267/34157051-ecdbd8ba-e4d0-11e7-9a84-bfd048282e75.png)

iamakulov commented:
I signed it!

iamakulov commented:
Also, a licensing question. In [`monitor-and-analyze.md`](https://github.com/iamakulov/WebFundamentals/blob/webpack-article/src/content/en/fundamentals/performance/webpack/monitor-and-analyze.md), I’m using a screen recording from the [webpack-bundle-analyzer](https://github.com/webpack-contrib/webpack-bundle-analyzer) repo. I’ve converted it from GIF to MP4 (i.e. it’s probably a modification). The repo is licensed under MIT.

I’ve attributed the video source in a `<figcaption>`. Would this be enough to comply with the license? (From my understanding, this should be, but just in case Google has stricter requirements.)

petele commented:
> For some reason, “Note:” blocks were recognized

`Note:` formatting looks for a specific regex, if it's not matched, it may not work. Our Dev server isn't perfect, but it should get it 95% of the time.

> Locally, {: .hide-from-toc } wasn’t hiding

Correct, it's only supported on production, not in our dev server

> Is there a way to enable diff-like formatting

No, sadly not. 

> I’d like to add “Next section” arrows

To add them, you need to add a `step_group: unique_id` to the `_toc.yaml` file for that section. It will show up in production, but isn't supported in the dev server.

Short version - our dev server simulates the real production environment, while it supports the majority of things that the production env supports, it doesn't support everything and it's always perfect.

iamakulov commented:
> > Is there a way to enable diff-like formatting
> No, sadly not.

Replaced with “Before”–“After” comments.

> > I’d like to add “Next section” arrows
> To add them, you need to add a `step_group: unique_id` to the `_toc.yaml` file for that section.

Added!

petele commented on src/content/en/fundamentals/performance/webpack/conclusion.md:
```diff
@@ -0,0 +1,36 @@
+Don’t forget to read [webpack docs](https://webpack.js.org/guides/) – they have plenty of other
+useful information.
+
+And make sure to play with the training app **TODO: link to apply these principles in practice!**
```
is there supposed to be a link to the TODO app here?

petele commented on src/content/en/fundamentals/performance/webpack/decrease-frontend-size.md:
```diff
@@ -0,0 +1,695 @@
+[EnvironmentPlugin](https://webpack.js.org/plugins/environment-plugin/). It works like the
+`DefinePlugin`, but reads the environment and replaces `process.env.` expressions automatically.
+
+Note: You’re not required to use the `UglifyJsPlugin`. You can use any different minifier as soon as
```
Three consecutive `notes` looks a little icky. Do they all need to be notes?

petele commented on src/content/en/fundamentals/performance/webpack/index.md:
```diff
@@ -0,0 +1,26 @@
+JavaScript.
+
+Note: We created a training app to play with optimizations described in this article. Try squeezing
+the most out of it to practice the tips **TODO: link**
```
TODO link?

petele commented on src/content/en/fundamentals/performance/webpack/monitor-and-analyze.md:
```diff
@@ -0,0 +1,252 @@
+the bundle). You’ll likely want to compare gzip sizes since that’s closer to what real users
+experience. Use the sidebar on the left to switch the sizes.
+
+Note: If you use the
```
same comment about consecutive notes

petele commented on src/content/en/fundamentals/performance/webpack/monitor-and-analyze.md:
```diff
@@ -0,0 +1,252 @@
+
+- **Large dependencies.** Why are they so large? Are there smaller alternatives (e.g., Preact
+instead of React)? Do you use all the code it includes (e.g., Moment.js includes a lot of locales
+that are often not used and could be dropped TODO: link to the repo with Moment.js optimizations)?
```
todo link?

petele commented:
@addyosmani - can you give an LGTM on this when it's ready to go please?

addyosmani commented:
Sure thing. Initial PR LGTM and I’ll take another read through once the other comments have been addressed.

On the comment about consecutive notes, I feel the content there is important - perhaps we make them less unsightly by converting one or two to regular text rather than a highlighted note? :)

iamakulov commented:
@petele @addyosmani Thanks for reviewing! :–)

I got rid of the consecutive notes. I had to remove the note about the webpack’s `-p` flag, but I didn’t really feel it fits well anyway. (Feel free to say if you still want to see in the article.)

The TODO links are not ready yet ­­­­– I’ll ping as soon as I add them.

petele commented:
@iamakulov - since todo links aren't ready yet, i assume i should wait to merge and push live until they're ready?

petele commented on src/content/en/fundamentals/performance/webpack/decrease-frontend-size.md:
```diff
@@ -0,0 +1,691 @@
+## Specify `NODE_ENV=production`
+
+Another way to decrease the front-end size is to set the `NODE_ENV`
+[environmental variable](https://superuser.com/questions/284342/what-are-path-
```
don't linebreak URLs, otherwise the link breaks. it's OK if this exceeds 80 chars

petele commented on src/content/en/fundamentals/performance/webpack/decrease-frontend-size.md:
```diff
@@ -0,0 +1,691 @@
+
+### Further reading {: .hide-from-toc }
+
+* [What “environment variables” are](https://superuser.com/questions/284342/
```
don't linebreak URLs, otherwise the link breaks. it's OK if this exceeds 80 chars

petele commented on src/content/en/fundamentals/performance/webpack/use-long-term-caching.md:
```diff
@@ -0,0 +1,647 @@
+
+* Webpack docs [about webpack runtime and manifest](https://webpack.js.org/concepts/manifest/)
+
+* [“Getting the most out of the CommonsChunkPlugin”](https://medium.com/webpack/
```
don't linebreak URLs, otherwise the link breaks. it's OK if this exceeds 80 chars

petele commented on src/content/en/fundamentals/performance/webpack/use-long-term-caching.md:
```diff
@@ -0,0 +1,647 @@
+
+* Webpack docs [about the CommonsChunkPlugin](https://webpack.js.org/plugins/commons-chunk-plugin/)
+
+* [“Getting the most out of the CommonsChunkPlugin”](https://medium.com/webpack/webpack-bits-
```
don't linebreak URLs, otherwise the link breaks. it's OK if this exceeds 80 chars. 

if you really wanted to line break this, you could do something like:

```
* [“Getting the most out of the
CommonsChunkPlugin”](https://medium.com/webpack/webpack-bits-more-and-more-and-more-and-more)
```

iamakulov replied:
> don't linebreak URLs, otherwise the link breaks

OK! (I checked locally, and they worked, but I’ll fix them.) Will fix this shortly.

> if you really wanted to line break this, you could do something like:

I’ve actually got test fails because of this: https://travis-ci.org/google/WebFundamentals/builds/325794409#L743. (If you see [the line which triggers the error](https://github.com/google/WebFundamentals/commit/3c3ac29abc1ce4380e9fcafabf9e325dcd85ed8f#diff-9f55b5ed9c2b6c50e93df9647ab49499R158), there’s a line break inside the link name.)

iamakulov commented:
> since todo links aren't ready yet, i assume i should wait to merge and push live until they're ready?

Yes, exactly! Thanks.

addyosmani commented:
The fixes for consecutive notes lgtm. It looks like we're still working on the line-break issues and TODOs here. Will take another read through once @iamakulov is happy with those have been addressed :)

petele commented:
@addyosmani - i've added the `DO NOT MERGE` tag, i'll ignore this until you remove it and let me know it's good to go!

addyosmani commented:
@iamakulov would it be possible to get an ETA on when the above remaining edits can be completed? :)

iamakulov commented:
@addyosmani I hate these delays, sorry so much for them :–(

Will fix line breaks on this Saturday. ETA on TODOs is the end of the next week (the preview should be earlier)

addyosmani commented:
@iamakulov Thanks! Took a look through the latest commit. I'll keep an eye out for any further updates to the TODOs.

iamakulov commented:
(JFYI: made a couple of minor changes)

addyosmani commented:
Awesome. Thanks for the updates!

iamakulov commented:
And... we’re done here!

(The links in the latest commit are offline right now, but they will become live as soon as we move the repos.)

addyosmani commented:
The repos for this project have been approved for launch. We're going to get them moved over to the GoogleChromeLabs org today and should be good to land this content at any time before/between. This should be fine as long as we're not tweeting the content out yet (I'd aim to get that done tomorrow).

iamakulov commented:
Added article links to https://github.com/GoogleChromeLabs/webpack-training-project.

I think we’re done.
